pmlc/rt/runtime_registry.h

```diff
@@ -17,56 +17,53 @@ namespace pmlc::rt {
 
 // Factory is used to instantiate a Runtime.
 //
-// Factory is invoked on-demand (typically well after process initialization) to
+// A Factory is typically a small, stateless function, safely instantiated at
+// static initialization time to defer Runtime instantiation to a time after
+// static initialization is complete.
+//
+// Typical runtime implementations have no need to explicitly use Factory.  To
+// register a DefaultConstructible Runtime at static initialization time, the
+// runtime implementation should use RuntimeRegistration, which will
+// automatically create and register a Factory.
+//
+// (Alternatively, if the Runtime is instantiated explicitly after static
+// initialization time, the caller may use bindRuntime() to add the runtime to
+// the system's global runtime map.)
+//
+// Factory is invoked on-demand (typically well after static initialization) to
 // instantiate a runtime, either as a specific request by the PlaidML caller or
 // for general device enumeration.  The registry is responsible for arranging
-// runtimes to be instantiated exactly once; Factory implementations do not need
-// to memoize their results.
+// runtimes to be instantiated exactly once; Factory implementations should not
+// memoize their results.
 using Factory = std::function<std::shared_ptr<Runtime>()>;
 
-// Loaders instantiate Factories.
-//
-// PlaidML Loaders are (typically) simple, stateless functions, automatically
-// registered via LoaderRegistration when their containing library is loaded
-// into the process.
-//
-// Loaders are invoked on-demand (typically well after process initialization)
-// to determine the runtime factories they're able to provide.
-using Loader = std::function<std::unordered_map<std::string, Factory>()>;
-
 namespace details {
 
-// Creates a global registration for a loader.
-void registerLoader(llvm::StringRef id, Loader loader);
+// Creates a global registration for a factory.
+void registerFactory(llvm::StringRef id, Factory factory);
```

Comment:
Can we make this part of the public interface? What's the rationale for `details` namespace?

Reply:
No problem; done.  (Details: because it's tricky to use correctly, because it requires that `initRuntimes()` is called after it, and because I don't think anything ever needs to use it; if you're post-main(), just instantiate your Runtime and call `registerRuntime()`.)